File: jlox/src/main/java/fr/ocus/lox/jlox/Scanner.java
package fr.ocus.lox.jlox;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static fr.ocus.lox.jlox.TokenType.AND;
import static fr.ocus.lox.jlox.TokenType.BANG;
import static fr.ocus.lox.jlox.TokenType.BANG_EQUAL;
import static fr.ocus.lox.jlox.TokenType.CLASS;
import static fr.ocus.lox.jlox.TokenType.COMMA;
import static fr.ocus.lox.jlox.TokenType.DOT;
import static fr.ocus.lox.jlox.TokenType.ELSE;
import static fr.ocus.lox.jlox.TokenType.EOF;
import static fr.ocus.lox.jlox.TokenType.EQUAL;
import static fr.ocus.lox.jlox.TokenType.EQUAL_EQUAL;
import static fr.ocus.lox.jlox.TokenType.FALSE;
import static fr.ocus.lox.jlox.TokenType.FOR;
import static fr.ocus.lox.jlox.TokenType.FUN;
import static fr.ocus.lox.jlox.TokenType.GREATER;
import static fr.ocus.lox.jlox.TokenType.GREATER_EQUAL;
import static fr.ocus.lox.jlox.TokenType.IDENTIFIER;
import static fr.ocus.lox.jlox.TokenType.IF;
import static fr.ocus.lox.jlox.TokenType.LEFT_BRACE;
import static fr.ocus.lox.jlox.TokenType.LEFT_PAREN;
import static fr.ocus.lox.jlox.TokenType.LESS;
import static fr.ocus.lox.jlox.TokenType.LESS_EQUAL;
import static fr.ocus.lox.jlox.TokenType.MINUS;
import static fr.ocus.lox.jlox.TokenType.NIL;
import static fr.ocus.lox.jlox.TokenType.NUMBER;
import static fr.ocus.lox.jlox.TokenType.OR;
import static fr.ocus.lox.jlox.TokenType.PLUS;
import static fr.ocus.lox.jlox.TokenType.PRINT;
import static fr.ocus.lox.jlox.TokenType.RETURN;
import static fr.ocus.lox.jlox.TokenType.RIGHT_BRACE;
import static fr.ocus.lox.jlox.TokenType.RIGHT_PAREN;
import static fr.ocus.lox.jlox.TokenType.SEMICOLON;
import static fr.ocus.lox.jlox.TokenType.SLASH;
import static fr.ocus.lox.jlox.TokenType.STAR;
import static fr.ocus.lox.jlox.TokenType.STRING;
import static fr.ocus.lox.jlox.TokenType.SUPER;
import static fr.ocus.lox.jlox.TokenType.THIS;
import static fr.ocus.lox.jlox.TokenType.TRUE;
import static fr.ocus.lox.jlox.TokenType.VAR;
import static fr.ocus.lox.jlox.TokenType.WHILE;

/**
 * @author Matthieu Honel <ocus51@gmail.com>
 * @since 2017-08-03
 */
class Scanner {
    private final String source;
    private final List<Token> tokens = new ArrayList<>();

    private int start = 0;
    private int current = 0;
    private int line = 1;

    Scanner(String source) {
        this.source = source;
    }

    List<Token> scanTokens() {
        while (!isAtEnd()) {
            start = current;
            scanToken();
        }

        tokens.add(new Token(EOF, "", null, line));
        return tokens;
    }

    private boolean isAtEnd() {
        return current >= source.length();
    }

    private void scanToken() {
        char c = advance();
        switch (c) {
            /* @formatter:off */
            case '(': addToken(LEFT_PAREN); break;
            case ')': addToken(RIGHT_PAREN); break;
            case '{': addToken(LEFT_BRACE); break;
            case '}': addToken(RIGHT_BRACE); break;
            case ',': addToken(COMMA); break;
            case '.': addToken(DOT); break;
            case '-': addToken(MINUS); break;
            case '+': addToken(PLUS); break;
            case ';': addToken(SEMICOLON); break;
            case '*': addToken(STAR); break;
            case '!': addToken(match('=') ? BANG_EQUAL : BANG); break;
            case '=': addToken(match('=') ? EQUAL_EQUAL : EQUAL); break;
            case '<': addToken(match('=') ? LESS_EQUAL : LESS); break;
            case '>': addToken(match('=') ? GREATER_EQUAL : GREATER); break;
            case '"': string(); break;
            /* @formatter:on */
            case '/':
                if (match('/')) {
                    while (peek() != '\n' && !isAtEnd()) advance();
                } else {
                    addToken(SLASH);
                }
                break;
            case ' ':
            case '\r':
            case '\t':
                break;
            case '\n':
                line++;
                break;
            default:
                if (isDigit(c)) {
                    number();
                } else if (isAlpha(c)) {
                    identifier();
                } else {
                    JLox.error(line, "Unexpected character.");
                }
                break;
        }
    }

    private boolean match(char expected) {
        if (isAtEnd()) return false;
        if (source.charAt(current) != expected) return false;
        current++;
        return true;
    }

    private boolean isDigit(char c) {
        return c >= '0' && c <= '9';
    }

    private boolean isAlpha(char c) {
        return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_';
    }

    private boolean isAlphaNumeric(char c) {
        return isAlpha(c) || isDigit(c);
    }

    private char advance() {
        current++;
        return source.charAt(current - 1);
    }

    private char peek() {
        if (isAtEnd()) return '\0';
        return source.charAt(current);
    }

    private char peekNext() {
        if (current + 1 >= source.length()) return '\0';
        return source.charAt(current + 1);
    }

    private void string() {
        while (peek() != '"' && !isAtEnd()) {
            if (peek() == '\n') line++;
            advance();
        }
        if (isAtEnd()) {
            JLox.error(line, "Unterminated string.");
            return;
        }
        advance(); // closing "

        String value = source.substring(start + 1, current - 1);
        addToken(STRING, value);
    }

    private void number() {
        while (isDigit(peek())) advance();

        if (peek() == '.' && isDigit(peekNext())) {
            advance();
            while (isDigit(peek())) advance();
        }

        addToken(NUMBER, Double.parseDouble(source.substring(start, current)));
    }

    private void identifier() {
        while (isAlphaNumeric(peek())) advance();

        // see if the identifier is reserved word.
        String text = source.substring(start, current);

        TokenType type = keywords.get(text);
        if (type == null) type = IDENTIFIER;
        addToken(type);
    }

    private void addToken(TokenType type) {
        addToken(type, null);

    }

    private void addToken(TokenType type, Object literal) {
        String text = source.substring(start, current);
        tokens.add(new Token(type, text, literal, line));
    }

    private static final Map<String, TokenType> keywords;

    static {
        keywords = new HashMap<>();
        /* @formatter:off */
        keywords.put("and",    AND);
        keywords.put("class",  CLASS);
        keywords.put("else",   ELSE);
        keywords.put("false",  FALSE);
        keywords.put("for",    FOR);
        keywords.put("fun",    FUN);
        keywords.put("if",     IF);
        keywords.put("nil",    NIL);
        keywords.put("or",     OR);
        keywords.put("print",  PRINT);
        keywords.put("return", RETURN);
        keywords.put("super",  SUPER);
        keywords.put("this",   THIS);
        keywords.put("true",   TRUE);
        keywords.put("var",    VAR);
        keywords.put("while",  WHILE);
        /* @formatter:on */
    }
}

File: pylox/pylox/scanner.py
from typing import List

from .interface import ScannerInterface
from .token import Token, TokenType


class Scanner(ScannerInterface):
    def __init__(self, source: str):
        self._source = source
        self._tokens = []  # type: List[Token]
        self._start = 0
        self._current = 0
        self._line = 1

    def scan_tokens(self) -> List[Token]:
        while not self._is_at_end():
            self._start = self._current
            self._scan_token()

        self._tokens.append(Token(TokenType.EOF, '', None, self._line))
        return self._tokens

    def _is_at_end(self):
        return self._current >= len(self._source)

    def _scan_token(self):
        c = self._advance()
        if c == '(':
            self._add_token(TokenType.LEFT_PAREN)
        elif c == ')':
            self._add_token(TokenType.RIGHT_PAREN)
        elif c == '{':
            self._add_token(TokenType.LEFT_BRACE)
        elif c == '}':
            self._add_token(TokenType.RIGHT_BRACE)
        elif c == ',':
            self._add_token(TokenType.COMMA)
        elif c == '.':
            self._add_token(TokenType.DOT)
        elif c == '-':
            self._add_token(TokenType.MINUS)
        elif c == '+':
            self._add_token(TokenType.PLUS)
        elif c == ';':
            self._add_token(TokenType.SEMICOLON)
        elif c == '*':
            self._add_token(TokenType.STAR)
        elif c == '!':
            self._add_token(TokenType.BANG_EQUAL if self._match('=') else TokenType.BANG)
        elif c == '=':
            self._add_token(TokenType.EQUAL_EQUAL if self._match('=') else TokenType.EQUAL)
        elif c == '<':
            self._add_token(TokenType.LESS_EQUAL if self._match('=') else TokenType.LESS)
        elif c == '>':
            self._add_token(TokenType.GREATER_EQUAL if self._match('=') else TokenType.GREATER)
        elif c == '/':
            if self._match('/'):
                while self._peek() != '\n' and not self._is_at_end():
                    self._advance()
            else:
                self._add_token(TokenType.SLASH)
        elif c in (' ', '\r', '\t'):
            pass  # ignore whitespace.
        elif c == "\n":
            self._line += 1  # incoming next line
        elif c == '"':
            self._string()
        else:
            if self._is_digit(c):
                self._number()
            elif self._is_alpha(c):
                self._identifier()
            else:
                from pylox import PyLox
                PyLox.error(self._line, 'Unexpected character.')

    def _add_token(self, token_type, literal=None):
        text = self._source[self._start:self._current]
        self._tokens.append(Token(token_type, text, literal, self._line))

    def _advance(self):
        self._current += 1
        return self._source[self._current - 1]

    def _match(self, expected):
        if self._is_at_end():
            return False
        if self._source[self._current] != expected:
            return False

        self._current += 1
        return True

    def _peek(self):
        if self._is_at_end():
            return '\0'
        return self._source[self._current]

    def _peek_next(self):
        if self._current + 1 >= len(self._source):
            return '\0'
        return self._source[self._current + 1]

    def _string(self):
        while self._peek() != '"' and not self._is_at_end():
            if self._peek() == '\n':
                self._line += 1
            self._advance()

        # Unterminated string
        if self._is_at_end():
            from pylox import PyLox
            PyLox.error(self._line, "Unterminated string.")

        # the closing "
        self._advance()

        value = self._source[self._start + 1: self._current - 1]
        self._add_token(TokenType.STRING, value)

    def _number(self):
        while self._is_digit(self._peek()):
            self._advance()

        if self._peek() == '.' and self._is_digit(self._peek_next()):
            self._advance()  # consume the "."
            while self._is_digit(self._peek()):
                self._advance()

        value = float(self._source[self._start:self._current])
        self._add_token(TokenType.NUMBER, value)

    def _identifier(self):
        while self._is_alpha_numeric(self._peek()):
            self._advance()

        # reserved word ?
        text = self._source[self._start:self._current]
        token_type = Scanner.KEYWORDS[text] if text in Scanner.KEYWORDS else TokenType.IDENTIFIER

        self._add_token(token_type)

    @staticmethod
    def _is_digit(c):
        return c.isdigit()

    @staticmethod
    def _is_alpha(c):
        return c.isalpha() or '_' == c

    def _is_alpha_numeric(self, c):
        return self._is_alpha(c) or self._is_digit(c)

    KEYWORDS = {
        'and': TokenType.AND,
        'class': TokenType.CLASS,
        'else': TokenType.ELSE,
        'false': TokenType.FALSE,
        'for': TokenType.FOR,
        'fun': TokenType.FUN,
        'if': TokenType.IF,
        'nil': TokenType.NIL,
        'or': TokenType.OR,
        'print': TokenType.PRINT,
        'return': TokenType.RETURN,
        'super': TokenType.SUPER,
        'this': TokenType.THIS,
        'true': TokenType.TRUE,
        'var': TokenType.VAR,
        'while': TokenType.WHILE,
    }
